### PR TITLE
decouple campaign price from asset state

### PR DESCRIPTION
**Pull request by evmbro**
removed the tokenPricePrecision property from assets

use priceDecimals rather than price precision (use 4, not 10**4)

provide priceDecimals as a parameter while creating campaign

provide priceDecimals as a parameter while updateing price on ApxRegistry (important for liquidations)

track priceDecimals when storing the highestSellPrice on asset state (important for liquidations)

make all the _token_value() methods look the same (operations ordering)

fix tests

add priceDecimalsPrecision to the campaign common state

fix deployer service scripts to take into account new priceDecimals parameter when creating campaigns

### CODE REVIEW

File: contracts/services/QueryService.sol [hunk 1]
```diff
@@ -517,12 +517,13 @@ contract QueryService is IQueryService {
         uint256 tokenAmount,
         IAssetCommon token,
         IToken stablecoin,
-        uint256 price
+        uint256 price,
+        uint256 priceDecimals
```
Comment (githinho) [inline contracts/services/QueryService.sol:521]: Why use `uint256`?!

File: contracts/managers/ACfManager.sol [hunk 2]
```diff
@@ -320,28 +347,42 @@ abstract contract ACfManager is IVersioned, IACfManager {
...
             state.asset,
             state.stablecoin
         );
-        return _token_value(tokenAmountForInvestment, state.tokenPrice, state.asset, state.stablecoin);
     }
 
     function _token_amount_for_investment(
         uint256 investment,
         uint256 tokenPrice,
+        uint256 tokenPriceDecimals,
```
Comment (githinho) [inline contracts/managers/ACfManager.sol:380]: Why use `uint256` for decimals?!

File: contracts/managers/ACfManager.sol [hunk 3]
```diff
@@ -140,16 +140,8 @@ abstract contract ACfManager is IVersioned, IACfManager {
...
-            (address treasury, uint256 fee) = _calculateFee();
-            if (fee > 0 && treasury != address(0)) {
-                sc.safeTransfer(treasury, fee);
-                sc.safeTransfer(msg.sender, fundsRaised - fee);
-            } else {
-                sc.safeTransfer(msg.sender, fundsRaised);
-            }
-        }
+        _safeFinalizeSale();
+        _safeDistributeFunds(msg.sender, fundsRaised, sc);
```
Comment (githinho) [inline contracts/managers/ACfManager.sol:144]: Why is a safe transfer needed?

File: contracts/services/QueryService.sol (hunk 1 above)
Comment (filipdujmusic) [reply contracts/services/QueryService.sol:521]: It's a read-only service so any optimization is pointless but I'll fix it in the new commit with other changes

File: contracts/managers/ACfManager.sol (hunk 3 above)
Comment (filipdujmusic) [reply contracts/managers/ACfManager.sol:144]: If for any reason FeeManager interaction fails we still want to allow the campaign manager to finalize the campaign , or else he's locked out of his funds. In other words, we want the fee calculation to silently fail

File: contracts/shared/Structs.sol [hunk 4]
```diff
@@ -70,7 +70,7 @@ contract Structs {
         address issuer;
         address paymentMethod;
```
Comment (mightymatth) [inline contracts/shared/Structs.sol:71]: Can we rename it to something like `paymentToken` or `paymentTokenAddress` or something else? 
as word **method** tells "how"  something is paid (e.g. cash, card, "u naturi").

File: contracts/shared/Structs.sol [hunk 5]
```diff
@@ -52,7 +52,7 @@ contract Structs {
         address issuerAddress;
         address paymentMethod;
```
Comment (mightymatth) [inline contracts/shared/Structs.sol:53]: same here

File: contracts/shared/Structs.sol (hunk 4 above)
Comment (filipdujmusic) [reply contracts/shared/Structs.sol]: done

**Review by githinho — APPROVED** [review]

**Comment by filipdujmusic** [comment]
thx !

AMPNET

Filip Dujmušić, COO

Phone: +385 95 891 5667

Email: ***@***.***

Web: https://ampnet.io


AMPnet IO d.o.o., Josipa Marohnica 3, 10000 Zagreb. VAT Registration:
HR57624879969.

Please consider the environment before printing this message.

**Review by mightymatth — APPROVED** [review]
Alles gut, but check whether it works in [frontend branch](https://github.com/AMPnet/amptzr/pull/96). demo is [here](https://amptzr-git-decouple-campaign-price-from-asset-state-ampnetx.vercel.app/).

Check all screens related to assets and campaigns. Try to create a new asset and campaign to see how it works. Keep the console opened to see possible errors. gl hf

**Pull request merged**